tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,434 @@
+                       "1" => 1,
+                       "0" => 0,
+                       "-1" => 1,
+                       null => null,
+                       not null => int.TryParse(samplingPriority, out var result) ? result : null
+                   };
         }
 
-        private static bool IsValidTraceId([NotNullWhen(true)] string? traceId)
+#if NETCOREAPP
+        private static int? SamplingPriorityToInt32(ReadOnlySpan<char> samplingPriority)
         {
-            if (string.IsNullOrEmpty(traceId))
+            if (samplingPriority.Equals("2", StringComparison.Ordinal))
             {
-                return false;
+                return 2;
             }
 
-            if (traceId!.Length != 32)
+            if (samplingPriority.Equals("1", StringComparison.Ordinal))
             {
+                return 1;
+            }
+
+            if (samplingPriority.Equals("0", StringComparison.Ordinal))
+            {
+                return 0;
+            }
+
+            if (samplingPriority.Equals("-1", StringComparison.Ordinal))
+            {
+                return -1;
+            }
+
+            return int.TryParse(samplingPriority, out var result) ? result : null;
+        }
+#endif
+
+        public bool TryExtract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter, out SpanContext? spanContext)
+            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        {
+            spanContext = null;
+
+            // get the "traceparent" header
+            var traceParentHeader = ParseUtility.ParseString(carrier, carrierGetter, TraceParentHeaderName)?.Trim();
+
+            if (string.IsNullOrEmpty(traceParentHeader) || !TryParseTraceParent(traceParentHeader!, out var traceParent))
+            {
+                // "traceparent" header is required
                 return false;
             }
 
+            var traceStateHeader = ParseUtility.ParseString(carrier, carrierGetter, TraceStateHeaderName)?.Trim();
+
+            if (string.IsNullOrEmpty(traceStateHeader) || !TryParseTraceState(traceStateHeader!, out var traceState))
+            {
+                // "tracestate" header is optional
+                traceState = default;
+            }
+
+            // if we can't get the more specific sampling priority from "tracestate",
+            // then fallback to the boolean in "traceparent"
+            var samplingPriority = traceState.SamplingPriority ??
+                                   (traceParent.Sampled ?
+                                        SamplingPriorityValues.AutoKeep :
+                                        SamplingPriorityValues.AutoReject);
+
+            spanContext = new SpanContext(
+                traceParent.TraceId,
+                traceParent.ParentId,
+                samplingPriority,
+                serviceName: null,
+                origin: traceState.Origin,
+                traceParent.RawTraceId,
+                traceParent.RawParentId);
```

Comment:
nit: I would add named arguments for these as they're all similar types, to avoid any chance of transposition during refactoring